d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java

```diff
@@ -531,6 +533,182 @@ public void done()
+
+    loadBalancerStrategyFactories.put("degrader", new DegraderLoadBalancerStrategyFactoryV3());
+    clientFactories.put(PropertyKeys.HTTP_SCHEME, new DoNothingClientFactory());
+
+    SimpleLoadBalancerState state =
+        new SimpleLoadBalancerState(executorService,
+            uriRegistry,
+            clusterRegistry,
+            serviceRegistry,
+            clientFactories,
+            loadBalancerStrategyFactories);
+
+    SimpleLoadBalancer loadBalancer =
+        new SimpleLoadBalancer(state, 5, TimeUnit.SECONDS, _d2Executor);
+
+    FutureCallback<None> balancerCallback = new FutureCallback<>();
+    loadBalancer.start(balancerCallback);
+    balancerCallback.get();
+
+    URI uri = URI.create("http://test.qd.com:5678");
+    Map<Integer, PartitionData> partitionData = new HashMap<>(1);
+    partitionData.put(DEFAULT_PARTITION_ID, new PartitionData(1d));
+    Map<URI, Map<Integer, PartitionData>> uriData = new HashMap<>(2);
+    uriData.put(uri, partitionData);
+
+    prioritizedSchemes.add(PropertyKeys.HTTP_SCHEME);
+
+    Set<URI> bannedSet = new HashSet<>();
+    clusterRegistry.put("cluster-1", new ClusterProperties("cluster-1", Collections.emptyList(),
+        Collections.emptyMap(), bannedSet, NullPartitionProperties.getInstance()));
+
+    serviceRegistry.put("foo", new ServiceProperties("foo",
+        "cluster-1",
+        "/foo",
+        Arrays.asList("degrader"),
+        Collections.<String,Object>emptyMap(),
+        null,
+        null,
+        prioritizedSchemes,
+        null));
+    uriRegistry.put("cluster-1", new UriProperties("cluster-1", uriData));
+
+    URI expectedUri = URI.create("http://test.qd.com:5678/foo");
+    URIRequest uriRequest = new URIRequest("d2://foo/52");
+
+    RequestContext serviceContext = new RequestContext();
+    serviceContext.putLocalAttr(AffinityRoutingURIProvider.AFFINITY_ROUTING_URI_PROVIDER, new AffinityRoutingURIProvider() {
+      private final Map<String, URI> uriMap = new HashMap<>();
+
+      @Override
+      public boolean isEnabled() {
+        return true;
+      }
+
+      @Override
+      public Optional<URI> getTargetHostURI(String clusterName) {
+        return Optional.ofNullable(uriMap.get(clusterName));
+      }
+
+      @Override
+      public void setTargetHostURI(String clusterName, URI targetHostURI) {
+        uriMap.put(clusterName, targetHostURI);
+      }
+    });
+
+    RewriteLoadBalancerClient client =
+        (RewriteLoadBalancerClient) loadBalancer.getClient(uriRequest, serviceContext);
+    Assert.assertEquals(client.getUri(), expectedUri);
+    AffinityRoutingURIProvider affinityRoutingURIProvider
+        = (AffinityRoutingURIProvider) serviceContext.getLocalAttr(AffinityRoutingURIProvider.AFFINITY_ROUTING_URI_PROVIDER);
+    Assert.assertEquals(affinityRoutingURIProvider.getTargetHostURI("cluster-1").get(), uri);
+  }
+
+  @Test
+  public void testOptimizedAffinityRoutingSkippedForTargetHostHint() throws Exception
```

Comment:
Please test the cases:
1) affinity routing provider is null. 
2) affinity routing provider is not enabled. 
3) request context target host is not set, and affinity routing provider is enabled and has a TargetHostURI set before loadBalancer.getClient.
4) request context target host is not set, and affinity routing provider is enabled and NOT has a TargetHostURI set before loadBalancer.getClient.

You could combine these two tests and use a data provider to provide vars that need different values, such as UriData, isAffinityRoutingURIProviderEnabled, affinityRoutingURIProviderTargetHostURI, requestContextTargetHost, affinityRoutingURIProviderTargetURI, expectedUri, etc.

Please add some comments before each test case in the data provider to describe what's tested for each case.

Reply:
1. All other test cases organically tests affinity routing provider  is null 
2. added
3. added
4. added